### PR TITLE
Move backup storage helpers into one file

### PR DESCRIPTION
**Pull request by stefan-burke**
## Summary
- move existing backup storage names, paths, freshness checks, and pruning into a focused module without changing behavior
- update all callers to use the new module directly
- split the matching tests into a focused file
- record two separate behavior changes raised in review as follow-up work

## Checks
- full pre-commit hook passed after each commit: lint, typecheck, duplication checks, copy checks, edge build, and coverage tests

## Risks
- no runtime behavior changes; backup naming, freshness checks, pruning error handling, archive creation, and restore behavior remain unchanged

## Summary by CodeRabbit

* **New Features**
  * Added dedicated backup storage handling for database-scoped backup locations, backup filename/path validation, recent-backup detection, and automatic pruning of old backups.
  * Updated the admin update workflow to use the new recent-backup check gate.

* **Bug Fixes**
  * Backup cleanup tolerates individual deletion failures without aborting the overall purge.

* **Documentation**
  * Added a follow-up note for backup storage edge cases and regression coverage.

* **Tests**
  * Added comprehensive new tests for backup storage helpers and adjusted existing backup-related test coverage to align with the change.

### CODE REVIEW

File: src/shared/db/backup-storage.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  const host = new URL(url).hostname;
+  const first = host.split(".")[0]!;
+
+  // Turso hostnames: {db-name}-{org}.turso.io — the full first segment is unique
+  if (host.endsWith(".turso.io")) return first;
+
+  // Bunny DB hostnames: {uuid}-{name}.lite.bunnydb.net — drop the UUID prefix
+  const dashIdx = first.indexOf("-");
+  if (dashIdx === -1) return first;
+  return first.slice(dashIdx + 1);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/backup-storage.ts:32]: Verified against `origin/main`: both behaviors predate this PR. This extraction is explicitly no-behavior-change, so changing supported URL/hostname handling here would violate its scope. Commit 053a934d records the `file:` namespace and non-Bunny hostname collision cases in `TODO.md`, including the proposed regression tests and starting point.

File: src/shared/db/backup-storage.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    stale.map(async (file) => {
+      try {
+        await deleteFile(file);
+        return file;
+      } catch {
+        return null;
+      }
+    }),
+  );
+  return compact(removed);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/backup-storage.ts:140]: No change here. Best-effort pruning is the existing documented contract on `main`: an external storage deletion failure must not make a successfully uploaded backup report failure. This PR was required to preserve that error handling exactly, including its test, so propagating the delete error would be a behavior regression rather than part of the extraction.

File: src/shared/db/backup-storage.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  const now = Date.now();
+  const files = await listFiles(backupDir(name));
+  for (const file of files) {
+    // Only real backups count — ignore anything else left in the folder, so a
+    // stray "{name}/manual-…Z.zip" can't spoof the freshness gate (mirrors
+    // pruneOldBackups).
+    if (!isBackupPath(file)) continue;
+    const ms = parseBackupTime(file);
+    if (ms !== null && now - ms < maxAgeMs) return true;
+  }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/backup-storage.ts:115]: Verified: the future-date behavior exists unchanged on `main`, and fixing it would alter the upgrade gate in a refactor required to preserve behavior. Commit 053a934d records this as a focused follow-up in `TODO.md`, including the clock-skew decision and future-timestamp regression test.

File: TODO.md [hunk 4]
```diff
@@ -1151,3 +1151,25 @@ Starting points: `src/features/api/payment-processing/completion.ts`,
...
+  `src/shared/db/backup-storage.ts` returns an empty name for a parseable local
+  `file:` URL and strips the first dashed part from non-Bunny hostnames. Decide
+  the supported URL schemes and hostnames, return a named local folder for local
+  URLs, and only remove Bunny DB's UUID prefix for `.lite.bunnydb.net`. Start
+  with tests for `file:database.db` and two distinct dashed HTTPS hostnames.
+- **Reject future-dated backups from the update gate.** `hasRecentBackup` in
+  `src/shared/db/backup-storage.ts` treats every future timestamp as recent
+  because its age is negative. Decide how much clock skew is acceptable, then
+  require a non-negative age (or a documented tolerance) before applying the
+  maximum age. Add a test with a future backup filename.
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply TODO.md:1175]: This is intentionally deferred because the task requires preserving `hasRecentBackup` exactly. The future-date behavior predates this PR on `main`; adding a non-negative-age rule or clock-skew tolerance would change the upgrade gate. The TODO entry records the decision and regression test needed for that separate behavior change.

File: TODO.md [hunk 5]
```diff
@@ -1151,3 +1151,25 @@ Starting points: `src/features/api/payment-processing/completion.ts`,
...
+PR #1837 only moves the existing backup storage helpers out of
+`src/shared/db/backup.ts`; it deliberately preserves their behavior. These
+possible behavior changes need separate decisions and regression tests:
+
+- **Keep every database namespace non-empty and distinct.** `dbName` in
+  `src/shared/db/backup-storage.ts` returns an empty name for a parseable local
+  `file:` URL and strips the first dashed part from non-Bunny hostnames. Decide
+  the supported URL schemes and hostnames, return a named local folder for local
+  URLs, and only remove Bunny DB's UUID prefix for `.lite.bunnydb.net`. Start
+  with tests for `file:database.db` and two distinct dashed HTTPS hostnames.
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply TODO.md:1170]: This is intentionally a follow-up rather than a change in this PR. The task requires an exact no-behavior-change extraction, and the same `dbName` behavior is already on `main`; changing accepted URL forms or backup namespaces here would mix a correctness change into the refactor. The TODO entry preserves the valid finding with its required regression cases for a separate change.
